chapter_natural-language-processing-pretraining/word-embedding-dataset.ipynb

```diff
@@ -0,0 +1,825 @@
+    "origin_pos": 23
+   },
+   "source": [
+    "### Negative Sampling\n",
+    "\n",
+    "We use negative sampling for approximate training. For a central and context word pair, we randomly sample $K$ noise words ($K=5$ in the experiment). According to the suggestion in the Word2vec paper, the noise word sampling probability $P(w)$ is the ratio of the word frequency of $w$ to the total word frequency raised to the power of 0.75 :cite:`Mikolov.Sutskever.Chen.ea.2013`.\n",
+    "\n",
+    "We first define a class to draw a candidate according to the sampling weights. It caches a 10,000 size random number bank.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public class RandomGenerator {\n",
+    "    /* Draw a random int in [0, n] according to n sampling weights. */\n",
+    "\n",
+    "    private List<Integer> population;\n",
+    "    private List<Double> samplingWeights;\n",
+    "    private List<Integer> candidates;\n",
+    "    private List<org.apache.commons.math3.util.Pair<Integer, Double>> pmf;\n",
+    "    private int i;\n",
+    "\n",
+    "    public RandomGenerator(List<Double> samplingWeights) {\n",
+    "        this.population =\n",
+    "                IntStream.range(0, samplingWeights.size()).boxed().collect(Collectors.toList());\n",
+    "        this.samplingWeights = samplingWeights;\n",
+    "        this.candidates = new ArrayList<>();\n",
+    "        this.i = 0;\n",
+    "\n",
+    "        this.pmf = new ArrayList<>();\n",
+    "        for (int i = 0; i < samplingWeights.size(); i++) {\n",
+    "            this.pmf.add(new org.apache.commons.math3.util.Pair(this.population.get(i), this.samplingWeights.get(i).doubleValue()));\n",
+    "        }\n",
+    "    }\n",
+    "\n",
+    "    public Integer draw() {\n",
+    "        if (this.i == this.candidates.size()) {\n",
+    "            this.candidates =\n",
+    "                    Arrays.asList((Integer[]) new EnumeratedDistribution(this.pmf).sample(10000, new Integer[] {}));\n",
+    "            this.i = 0;\n",
+    "        }\n",
+    "        this.i += 1;\n",
+    "        return this.candidates.get(this.i - 1);\n",
+    "    }\n",
+    "}\n",
+    "\n",
+    "RandomGenerator generator =\n",
+    "        new RandomGenerator(Arrays.asList(new Double[] {2.0, 3.0, 4.0}));\n",
+    "Integer[] generatorOutput = new Integer[10];\n",
+    "for (int i = 0; i < 10; i++) {\n",
+    "    generatorOutput[i] = generator.draw();\n",
+    "}\n",
+    "System.out.println(Arrays.toString(generatorOutput));"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public static ArrayList<ArrayList<Integer>> getNegatives(\n",
+    "        ArrayList<ArrayList<Integer>> allContexts, Integer[][] corpus, int K) {\n",
+    "    LinkedHashMap<Object, Integer> counter = Vocab.countCorpus2D(corpus);\n",
```

Comment:
can the Object type be more specific? like String

Reply:
So yes, originally it was set as a String but it isn't always the case where key is a String. There was a case I encountered where it was Integers. What approach should I take in this case? I just thought I could maybe try to use generics... what do you think?

Reply:
Ok if that's the case, I recommend to use ? LinkedHashMap<?, Integer>

Reply:
Yes this is what I was looking for, thanks.

Done